third_party/2/concurrent/futures/__init__.pyi

```diff
@@ -8,12 +12,12 @@ class Future(Generic[_T]):
     def running(self) -> bool: ...
     def done(self) -> bool: ...
     def result(self, timeout: float = ...) -> _T: ...
-    def exception(self, timeout: float = ...) -> BaseException: ...
+    def exception(self, timeout: float = ...) -> _ExceptionType: ...
```

Comment:
I don't quite follow your arguments against the `Union`.

The `exception()` method can return any arbitrary old-style object, so it's correct to say that callers can only safely use attributes that are present on both old-style objects and `BaseException`.  The `Union` helps enforce this restriction.